azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -300,28 +288,43 @@ public VMImage getImage(final String id) {
 
    @Override
    public Iterable<Location> listLocations() {
+      final Iterable<String> whiteListZoneName = findWhiteListOfRegions();
+
+      final Iterable<String> vmLocations = FluentIterable.from(api.getResourceProviderApi().get("Microsoft.Compute"))
+              .filter(new Predicate<ResourceProviderMetaData>() {
+                 @Override
+                 public boolean apply(ResourceProviderMetaData input) {
+                    return input.resourceType().equals("virtualMachines");
+                 }
+              })
+              .transformAndConcat(new Function<ResourceProviderMetaData, Iterable<String>>() {
+                 @Override
+                 public Iterable<String> apply(ResourceProviderMetaData resourceProviderMetaData) {
+                    return resourceProviderMetaData.locations();
+                 }
+              });
+
+      List<Location> locations = FluentIterable.from(api.getLocationApi().list())
+              .filter(new Predicate<Location>() {
+                 @Override
+                 public boolean apply(Location location) {
+                    return Iterables.contains(vmLocations, location.displayName());
+                 }
+              })
+              .filter(new Predicate<Location>() {
+                 @Override
+                 public boolean apply(Location location) {
+                    return whiteListZoneName == null ? true : Iterables.contains(whiteListZoneName, location.name());
+                 }
+              })
+              .toList();
+
+      return locations;
+   }
 
-      List<Location> locations = api.getLocationApi().list();
-
-      List<ResourceProviderMetaData> resources = api.getResourceProviderApi().get("Microsoft.Compute");
-
-      final List<String> vmLocations = new ArrayList<String>();
-
-      for (ResourceProviderMetaData m : resources){
-         if (m.resourceType().equals("virtualMachines")){
-            vmLocations.addAll(m.locations());
-            break;
-         }
-      }
-
-      Iterable<Location> result = Iterables.filter(locations, new Predicate<Location>() {
-         @Override
-         public boolean apply(Location input) {
-            return vmLocations.contains(input.displayName());
-         }
-      });
-
-      return result;
+   private Iterable<String> findWhiteListOfRegions() {
+      if (providerMetadata.getDefaultProperties().get(LocationConstants.PROPERTY_REGIONS) == null)  return null;
+      return Splitter.on(",").trimResults().split((CharSequence) providerMetadata.getDefaultProperties().get(LocationConstants.PROPERTY_REGIONS));
    }
 
    private String getResourceGroupFromId(String id) {
```

Comment:
If we have a global property that filters locations, we should also filter the `listNodes` operation.

How should the `getImage` and `getNode` behave when invoked with a resource that is from another location?